src/chain/store/chain_store.rs

```diff
@@ -594,6 +590,82 @@ where
 
         Ok(digest)
     }
+
+    /// Get the [`TipsetKeys`] for a given epoch. The returned key will never be null.
+    pub fn get_epoch_tsk(
+        &self,
+        tipset: Arc<Tipset>,
+        round: ChainEpoch,
+    ) -> Result<TipsetKeys, Error> {
+        let ts = self
+            .tipset_by_height(round, tipset, true)
+            .map_err(|e| Error::Other(format!("Could not get tipset by height {e:?}")))?;
+        Ok(ts.key().clone())
+    }
+
+    /// Gets look-back tipset for block validations.
+    ///
+    /// The look-back tipset for a round is the tipset with epoch `round -
+    /// chain_finality`. Chain finality is usually 900. The 'heaviest_tipset' is
+    /// a reference point in the blockchain. It must be a child of the tipset
+    /// we're searching for.
+    pub fn get_lookback_tipset_for_round(
+        self: &Arc<Self>,
+        chain_config: Arc<ChainConfig>,
+        heaviest_tipset: Arc<Tipset>,
+        round: ChainEpoch,
+    ) -> Result<(Arc<Tipset>, Cid), Error>
+    where
+        DB: Send + Sync + 'static,
+    {
+        let version = chain_config.network_version(round);
+        let lb = if version <= NetworkVersion::V3 {
+            ChainEpoch::from(10)
+        } else {
+            chain_config.policy.chain_finality
+        };
+        let lbr = (round - lb).max(0);
+
+        // More null blocks than lookback
+        if lbr >= heaviest_tipset.epoch() {
+            // This situation is extremely rare so it's fine to compute the
+            // state-root without caching.
+            let genesis_timestamp = self.genesis().map_err(anyhow::Error::from)?.timestamp();
+            let beacon = Arc::new(chain_config.get_beacon_schedule(genesis_timestamp));
+            let (state, _) = crate::state_manager::apply_block_messages(
+                Arc::clone(self),
+                Arc::clone(&chain_config),
+                beacon,
+                Arc::new(crate::interpreter::RewardActorMessageCalc),
+                // Creating new WASM engines is expensive (takes seconds to
+                // minutes). It's only acceptable here about this situation is
```

Comment:
```suggestion
                // minutes). It's only acceptable here because this situation is
```

Reply:
Heh, my brain must have blanked out for a second there.